src/main/java/org/jabref/gui/externalfiles/LinkedFileAutoRenamer.java

```diff
@@ -0,0 +1,111 @@
+import org.slf4j.LoggerFactory;
+
+public class LinkedFileAutoRenamer {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LinkedFileAutoRenamer.class);
+
+    private final GuiPreferences preferences;
+    private final BibDatabaseContext bibDatabaseContext;
+
+    public LinkedFileAutoRenamer(BibDatabaseContext bibDatabaseContext, GuiPreferences preferences) {
+        this.bibDatabaseContext = bibDatabaseContext;
+        this.preferences = preferences;
+        bibDatabaseContext.getDatabase().registerListener(this);
+    }
+
+    @Subscribe
+    public void listen(FieldChangedEvent event) {
+        if (!preferences.shouldAutoRenameLinkedFiles()) {
+            return;
+        }
+
+        BibEntry entry = event.getBibEntry();
+        Field changedField = event.getField();
+
+        String fileNamePattern = preferences.getFilePreferences().getFileNamePattern();
+        if (fileNamePattern == null || fileNamePattern.isBlank()) {
+            return;
+        }
+
+        Set<String> patternFields = new HashSet<>();
+
+        BracketedPattern.expandBrackets(fileNamePattern, bracket -> {
+            List<String> fieldParts = BracketedPattern.parseFieldAndModifiers(bracket);
+            String fieldName = fieldParts.getFirst();
+            patternFields.add(fieldName);
+            return "";
+        });
+
+        LOGGER.debug("Parsed pattern fields: {}", patternFields);
+
+        boolean relevantFieldChanged = false;
+
+        if (patternFields.contains("bibtexkey") && changedField.equals(InternalField.KEY_FIELD)) {
+            relevantFieldChanged = true;
+            LOGGER.debug("Changed field '{}' is part of the pattern.", changedField.getName());
+        }
+
+        if (patternFields.contains(changedField.getName())) {
+            relevantFieldChanged = true;
+            LOGGER.debug("Changed field '{}' is part of the pattern.", changedField.getName());
+        }
+
+        if (!relevantFieldChanged) {
+            LOGGER.debug("Changed field '{}' is not relevant for renaming.", changedField.getName());
+            return;
+        }
+
+        handleEntryChange(entry, fileNamePattern);
+    }
+
+    private void handleEntryChange(BibEntry entry, String fileNamePattern) {
+        List<Path> fileDirectories = bibDatabaseContext.getFileDirectories(preferences.getFilePreferences());
+
+        LOGGER.debug("Handling entry change for entry: {}", entry.getCitationKey().orElse("unknown"));
+
+        List<LinkedFile> linkedFiles = entry.getFiles();
+
+        for (LinkedFile linkedFile : linkedFiles) {
+            try {
+                String newFileName = FileUtil.createFileNameFromPattern(bibDatabaseContext.getDatabase(), entry, fileNamePattern);
+                LOGGER.debug("Generated new file name: '{}'", newFileName);
+
+                boolean renamed = FileUtil.renameLinkedFileToName(linkedFile, entry, newFileName, false, bibDatabaseContext, fileDirectories, preferences.getFilePreferences());
```

Comment:
Passing the boolean `false` like this is against best practices, please define a constant OVEWRITE_EXISTING_FILE = true, and pass `!OVEWRITE_EXISTING_FILE` to the method.